hal/armv7a/zynq7000/interrupts.c

```diff
@@ -19,52 +19,225 @@
+{
+	u32 mask = *(interrupts_common.gic + dipr0 + (irqn >> 2)) & ~(0xff << ((irqn & 0x3) << 3));
+
+	*(interrupts_common.gic + dipr0 + (irqn >> 2)) = mask | ((priority & 0xff) << ((irqn & 0x3) << 3));
+}
+
+
+static inline u32 interrupts_getPriority(unsigned int irqn)
+{
+	return (*(interrupts_common.gic + dipr0 + (irqn >> 2)) >> ((irqn & 0x3) << 3)) & 0xff;
 }
 
 
 int hal_interruptsSetHandler(intr_handler_t *h)
 {
+	spinlock_ctx_t sc;
+
+	if (h == NULL || h->f == NULL || h->n >= SIZE_INTERRUPTS)
+		return -1;
+
+	hal_spinlockSet(&interrupts_common.spinlock[h->n], &sc);
+	_intr_add(&interrupts_common.handlers[h->n], h);
+
+	interrupts_setPriority(h->n, 0xa);
+	interrupts_setCPU(h->n, 0x1);
+	interrupts_enableIRQ(h->n);
+
+	hal_spinlockClear(&interrupts_common.spinlock[h->n], &sc);
+
 	return 0;
 }
 
 
 char *hal_interruptsFeatures(char *features, unsigned int len)
 {
-	return 0;
+	hal_strncpy(features, "Using gic interrupt controller", len);
+	features[len - 1] = 0;
+
+	return features;
 }
 
 
 int hal_interruptsDeleteHandler(intr_handler_t *h)
 {
+	spinlock_ctx_t sc;
+
+	if (h == NULL || h->f == NULL || h->n >= SIZE_INTERRUPTS)
+		return -1;
+
+	hal_spinlockSet(&interrupts_common.spinlock[h->n], &sc);
+	_intr_remove(&interrupts_common.handlers[h->n], h);
+
+	if (interrupts_common.handlers[h->n] == NULL)
+		interrupts_disableIRQ(h->n);
+
+	hal_spinlockClear(&interrupts_common.spinlock[h->n], &sc);
+
 	return 0;
 }
 
 
 /* Function initializes interrupt handling */
 void _hal_interruptsInit(void)
 {
+	u32 i;
+
+	for (i = 0; i < SIZE_INTERRUPTS; ++i) {
+		interrupts_common.handlers[i] = NULL;
+		interrupts_common.counters[i] = 0;
+		hal_spinlockCreate(&interrupts_common.spinlock[i], "interrupts");
+	}
+
+	interrupts_common.gic = (void *)(((u32)&_end + (5 * SIZE_PAGE) - 1) & ~(SIZE_PAGE - 1));
+
+	/* Initialize Distributor of the gic
+	 * enable_secure = 0 */
+	*(interrupts_common.gic + ddcr) &= ~0x3;
+
+	/* Set default priorities - 10 for the SGI (IRQID: 0 - 15), PPI (IRQID: 16 - 31), SPI (IRQID: 32 - 95) */
+	for (i = 0; i < SIZE_INTERRUPTS; ++i)
+		interrupts_setPriority(i, 0xa);
+
+	/* Set required configuration and CPU_0 as a default processor */
+	for (i = SPI_FIRST_IRQID; i < SIZE_INTERRUPTS; ++i) {
+		interrupts_setConf(i, spiConf[i - SPI_FIRST_IRQID]);
+		interrupts_setCPU(i, 0x1);
+	}
+
+	/* Disable interrupts */
+	*(interrupts_common.gic + dicer0) = 0xffffffff;
+	*(interrupts_common.gic + dicer0 + 1) = 0xffffffff;
+	*(interrupts_common.gic + dicer0 + 2) = 0xffffffff;
+
+	/* enable_secure = 1 */
+	*(interrupts_common.gic + ddcr) |= 0x3;
+
+	*(interrupts_common.gic + cicr) &= ~0x3;
+
+	/* Initialize CPU Interface of the gic
+	 * set the maximum priority mask */
+	*(interrupts_common.gic + cpmr) = (*(interrupts_common.gic + cpmr) & ~0x1f) | 0x1f;
 
+	/* EnableS = 1; EnableNS = 1; AckCtl = 1; FIQEn = 0 */
+	*(interrupts_common.gic + cicr) = (*(interrupts_common.gic + cicr) & ~0x7) | 0x7;
```

Comment:
You're setting these bits anyway, no need to clear them

```suggestion
	*(interrupts_common.gic + cpmr) |= 0x1f;

	/* EnableS = 1; EnableNS = 1; AckCtl = 1; FIQEn = 0 */
	*(interrupts_common.gic + cicr) |= 0x7;
```

Reply:
Done